### PR TITLE
Add metrics around batch creation and eventually received replies

### PR DESCRIPTION
**Pull request by igor-aptos**

### CODE REVIEW

**Review by zekun000 — APPROVED** [review]

File: consensus/src/quorum_store/proof_coordinator.rs [hunk 1]
```diff
@@ -67,40 +74,78 @@ impl IncrementalProofState {
...
     fn ready(&self, validator_verifier: &ValidatorVerifier, my_peer_id: PeerId) -> bool {
-        self.aggregated_signature.contains_key(&my_peer_id)
-            && validator_verifier
-                .check_voting_power(self.aggregated_signature.keys())
-                .is_ok()
+        if self.aggregated_signature.contains_key(&my_peer_id)
+            && self.aggregated_voting_power >= validator_verifier.quorum_voting_power()
+        {
+            let recheck = validator_verifier.check_voting_power(self.aggregated_signature.keys());
+            if recheck.is_err() {
```
Comment (sitalkedia) [inline consensus/src/quorum_store/proof_coordinator.rs:110]: Should we panic in this case?

Comment (igor-aptos) [reply consensus/src/quorum_store/proof_coordinator.rs:110]: what is the point in the panic? node can continue operating fine, will collect a few more votes and pass.

we will have alert on error! to actually fix whathever is causing the issue.

that is the main distinction between error!() and panic!(), both should never happen and require code fix, but panic should be used when node doesn't know how to proceed

Comment (sitalkedia) [reply consensus/src/quorum_store/proof_coordinator.rs:110]: If this is an invariant we are trying to maintain then it's better to panic! right? In this case, something has gone really wrong - may be node's internal state is corrupt and it shouldn't proceed further.

File: consensus/src/quorum_store/proof_coordinator.rs [hunk 2]
```diff
@@ -143,35 +185,36 @@ impl ProofCoordinator {
...
+                return Ok(Some(proof));
+            }
         }
         Ok(None)
     }
 
     async fn expire(&mut self) {
         let mut batch_ids = vec![];
         for signed_digest_info in self.timeouts.expire() {
-            counters::TIMEOUT_BATCHES_COUNT.inc();
```
Comment (zekun000) [inline consensus/src/quorum_store/proof_coordinator.rs:172]: missing the counter?

**Review by zekun000 — APPROVED** [review]

File: consensus/src/quorum_store/proof_coordinator.rs (hunk 2 above)
Comment (igor-aptos) [reply consensus/src/quorum_store/proof_coordinator.rs:172]: I'll add it back, but it is duplicate with counters::BATCH_SUCCESSFUL_CREATION

File: consensus/src/quorum_store/proof_coordinator.rs (hunk 1 above)
Comment (igor-aptos) [reply consensus/src/quorum_store/proof_coordinator.rs:99]: I think it is fine to alert and continue, but if you feel strongly I can panic too.

**Review by sitalkedia — APPROVED** [review]

File: consensus/src/quorum_store/proof_coordinator.rs (hunk 1 above)
Comment (sitalkedia) [reply consensus/src/quorum_store/proof_coordinator.rs:99]: IMO it makes sense to panic here but will leave it to you.

File: consensus/src/quorum_store/proof_coordinator.rs [hunk 3]
```diff
@@ -35,17 +35,25 @@ pub(crate) enum ProofCoordinatorCommand {
...
+    aggregated_voting_power: u128,
+    completed: bool,
 }
 
 impl IncrementalProofState {
     fn new(info: SignedDigestInfo) -> Self {
         Self {
             info,
             aggregated_signature: BTreeMap::new(),
+            aggregated_voting_power: 0,
```
Comment (gelash) [inline consensus/src/quorum_store/proof_coordinator.rs:47]: I think it makes sense now to create an SignatureAggregator utility, maybe even put it somewhere with ValidatorVerifier.rs and would also allow testing properly. Can return keys, do the linear algorithm with aggregated voting power/completed / recheck and everything, and would be easy to unit test too.

And actually having such tests, even basic ones to start, would be great?

**Pull request merged**